docs/ConfigHelp.md
# File Naming Convention and Location #
SuperSID uses the parameters set in the provided configuration file passed as argument on the command line. By convention, the file name is "supersid.cfg".
It is possible to have more than one configuration file to launch the SuperSID application with different parameters. For example to specify different list of Stations to monitor or to choose between graphic or text mode interface (like 'supersid.text.cfg' and 'supersid.wx.cfg').

The file can be located in any accessible directory providing that the full qualified path is given. On the other hand and for previous version compatibility, the configuration file argument can be omitted. In this case, the default "supersid.cfg" is searched:
  1. In the '../Config/..' folder
  2. In the local directory
  3. In the home directory '~/' (*nix system only)
 
# File Organization #
The configuration file is a simple text file formatted as a classic '.ini' structure i.e. sections with squared brackets and a list of pairs 'key=value"
 
 The supported sections are:
  * [PARAMETERS](#id-section1)
  * [STATION_x](#id-section2) where x is a number in [1..n]
  * [Capture](#id-section3)
  * [Email](#id-section4)
  * [FTP](#id-section5)
  
<div id='id-section1'/>

## [PARAMETERS] ##
This section groups most of the parameters identifying your SuperSID monitor. Some optional parameters offer the possibility to change some default values used by the program.

### Monitor Identification ###
  * contact: email or phone number of the SuperSID owner. *Mandatory*
  * site_name: unique identification of the SuperSID monitor. *Mandatory*
  * monitor_id: unique id to distinguish the monitors running on one site
  * longitude: in decimal form
  * latitude: in decimal form
  * utc_offset:
  * time_zone:
  
### Log Parameters ###
  * audio_sampling_rate: **48000** or **96000** (you can experiment with other value as long as your card support them)
  * log_interval: number of second between two reading. Default is '**5**' seconds. Reading/sound capture last one second.
  * log_type: **filtered** or **raw**. When **filtered** is indicated, *bema_wing* function is called to smoothen raw data before writting the file else in **raw** mode, captured data are written 'as is'. Note that *sidfile.py* can be used as an utility to apply 'bema_wing' function to an existing file (raw or not) to smoothen its data.
  * data_path: fully qualified path where files will be written. If not mentionned then '../Data/' is used.
  * log_format:
    - **sid_format**: one file per station with first data column as timestamp and second data column as captured value
    - **supersid_format**: one file for all station. No timestamp but one data column per station. Each line is *log_interval* seconds after the previous, first line at 0:00:00UTC.
    - **supersid_extended**: one file for all station. First data column is extended timestamp HH:MM:SS.mmmmm and following data column as one per station.
  * hourly_save: **yes** / **no** (default). If **yes** then a raw file is written every hour to limit data loss.
  
### FTP to Standford server ###
Version 1.4: FTP information are no longer part of the [PARAMETERS] section. Refer to the [FTP] section below.
  
### Extra ###
  * scaling_factor:
  * mode: [ignored] **Server**, **Client**, **Standalone** (default) . Reserved for future client/server dev.
  * viewer: **text** for text mode light interface, **wx** for *wxPython* GUI or **tk** for TkInter GUI (default)
  * bema_wing: beta_wing parameter for sidfile.filter_buffer() calculation. Default is '**6**'.

  * number_of_stations: specify the number of stations to monitor. Each station is described within its own section.

<div id='id-section2'/>

## [STATION_x] ##
Each station to monitor is enumerated from 1 till n=*number_of_stations*. For each station, one must provide:
  * call_sign: Station ID (various VLF station lists exist like [AAVSO's] (http://www.aavso.org/vlf-station-list) and [Wikipedia's] (http://en.wikipedia.org/wiki/Very_low_frequency#List_of_VLF_transmissions))
  * frequency: emission frequency in Hz
  * color: [rgbyw] to draw multiple graph together in *SuperSID_plot.py*.
  
<div id='id-section3'/>

## [Capture] ##
This section can be omitted if you plan to use the 'pyaudio' library. If you want to use the "alsaaudio" library then you can declare:
  * Audio: python library to use **alsaaudio** or **pyaudio** (default), **server** reserved for client/server future dev.
  * Card: [for alsaaudio only] card name for capture. Default is 'External'.
  * PeriodSize: [for alsaaudio only] period size for capture. Default is '128'.
  
<div id='id-section4'/>

## [Email] ##
The 'supersid_plot.py' program can send you an email with the attached plot as a PDF file. In order to use this feature, you must provide the information necessary to contact your email server as well as which email to use.
  * from_mail: sender's emai
  * to_mail: recipient's email
  * email_server: email server to use (SMPT)
  * email_login: [optional] if your server requires a login for identification
  * email_password: [optional] if your server requires a password for identification
  
<div id='id-section5'/>

## [FTP] ##
Group all parameters to send data to an FTP server i.e. Standford data repository.
  * automatic_upload: [yes/no] if set to 'yes' then trigger the FTP data upload
  * ftp_server: URL of the server (sid-ftp.stanford.edu)
  * ftp_directory: target folder on the FTP server where files should be written (on Standford's server: /incoming/SuperSID/NEW/)
  * local_tmp: local temporary directpry used to write the files before their upload
  * call_signs: list of recorded stations to upload. Not all recorded stations might be of interrest: list only the most relevant one(s).

supersid/config.py
#!/usr/bin/python
"""Config parses a supersid's .cfg file

Parameter access: all keys are forced to lowercase
  - for parameters: config['site_name'], config['longitude'], etc...
  - for stations:   config.stations[i] is a triplet: (call_sign, frequency, color)

Note: len(config.stations) == config['number_of_stations'] - sanity check -
"""
#
# Eric Gibert
#
#   Change Log:
#   20140816:
#   - define CONSTANTS to ensure universal usage
#   20150801:
#   - add the [FTP] section
#
from __future__ import print_function   # use the new Python 3 'print' function
import os.path
try:
    import ConfigParser
except ImportError:
    import configparser as ConfigParser

# constant for log_type
FILTERED, RAW = 'filtered', 'raw'
# constant for station parameters
CALL_SIGN, FREQUENCY, COLOR = 'call_sign', 'frequency', 'color'
# constant for log_format
SID_FORMAT, SUPERSID_FORMAT = 'sid_format', 'supersid_format'
SUPERSID_EXTENDED, BOTH_EXTENDED = 'supersid_extended', 'both_extended' # with 5 decimals timestamp
# Default value for alsaaudio Device parameter. Should be something the
# alsaaudio library would never accept.
DEVICE_DEFAULT = 'FOOBAR1234xyz'

class Config(dict):
    """Dictionary containing the key/values pair read from a .cfg file"""
    CONFIG_PATH_NAME = "../Config/" # default for historical reasons
    DATA_PATH_NAME   = "../Data/"   # default for historical reasons - can be overwritten by 'data_path'

    def __init__(self, filename="supersid.cfg"):
        """Read the given .cfg file (formatted as a .ini windows file) or tries to find one.
           All its key/values pairs are stored as a dictionary (self)
        :param filename: superSID .cfg file
        :return: nothing
        """
        self.version = "1.4 20150801"
        dict.__init__(self)         # Config objects are dictionaries
        self.config_ok = True       # Parsing success/failure
        self.config_err = ""        # Parsing failure error message
        config_parser = ConfigParser.ConfigParser()

        if filename == "supersid.cfg": # let's look in various places
            self.filenames = config_parser.read([Config.CONFIG_PATH_NAME + filename,   # historic location
                                                 os.path.join('.', filename),            # current location
                                                 os.path.expanduser(os.path.join('~', filename))]) # $home in *nix
        else: # a specific file has been given
            self.filenames = config_parser.read(filename)

        if len(self.filenames) == 0:
            self.config_ok = False
            self.config_err = "Cannot find configuration file: " + filename
            return

        # each section (dico entry) matches a list of parameters
        # each parameter has a key description, a type for cast, a default value or None if mandatory
        sections = { 'PARAMETERS': ( # optional entries
                                    ('contact', str, None),             # email of the SuperSID owner
                                    ('hourly_save', str, "no"),         # new flag: yes/no to save every hours
                                    ('data_path', str, ""),             # new: to override DATA_PATH_NAME by user
                                    ('log_format', str, SID_FORMAT),    # sid_format (default), supersid_format
                                    ('mode', str, 'Standalone'),        # Server, Client, Standalone (default)
                                    ('viewer', str, 'tk'),              # text, wx, tk (default)
                                    ('bema_wing', int, 6),              # beta_wing for sidfile.filter_buffer()
                                    # mandatory entries
                                    ('site_name', str, None),
                                    ('longitude', str, None),
                                    ('latitude', str, None),
                                    ('utc_offset', str, None),
                                    ('time_zone',  str, None),
                                    ('monitor_id', str, None),
                                    ('log_type',  str, None),           # 'filtered' or 'raw'

                                    ('audio_sampling_rate', int, None),
                                    ('log_interval', int, None),
                                    ('number_of_stations', int, None),
                                    ('scaling_factor', float, None)
                                    ),

                      "Capture":   (("Audio", str, 'pyaudio'),          # soundcard: alsaaudio or pyaudio ; server
                                    ("Card", str, 'External'),          # alsaaudio: card name for capture
                                    ("Device", str, DEVICE_DEFAULT),         # USe instead of card for alsaaudio
                                    ("PeriodSize", int, 128)            # alsaaudio: period size for capture
                                    ),

                      "Linux":     (("Audio", str, 'pyaudio'),          # soundcard: alsaaudio or pyaudio ; server
                                    ("Card", str, 'External'),          # alsaaudio: card name for capture
                                    ("PeriodSize", int, 128)            # alsaaudio: period size for capture
                                    ),

                      "Email":     (("from_mail", str, ""),             # sender email
                                    ("to_mail", str, ""),               # recipient email
                                    ("email_server", str, ""),          # your email server (SMPT)
                                    ("email_port", str, ""),            # your email server's port (SMPT)
                                    ("email_login", str, ""),           # if your server requires a login
                                    ("email_password", str, "")         # if your server requires a passwrd
                                    ),
                      "FTP":       (('automatic_upload',  str, "no"),   # yes/no: to upload the file to the remote FTP server
                                    ('ftp_server',  str, ""),           # address of the server like sid-ftp.stanford.edu
                                    ('ftp_directory', str, ""),         # remote target directory to write the files
                                    ('local_tmp', str, ""),             # local tmp folder to generate files before upload
                                    ('call_signs', str, "")             # list of stations to upload (sub-set of [stations])
                                    ),
                    }

        self.sectionfound = set()
        for section, fields in sections.items():
            # go thru all the current section's fields
            for pkey, pcast, pdefault in fields:
                try:
                    self[pkey] = pcast(config_parser.get(section, pkey))
                except ValueError:
                    self.config_ok = False
                    self.config_err = "'%s' is not of the type %s in 'supersid.cfg'. Please check." % (pkey, pcast)
                    return
                except ConfigParser.NoSectionError:
                    #it's ok: some sections are optional
                    pass
                except ConfigParser.NoOptionError:
                    if pdefault is None: # missing mandatory parameter
                        self.config_ok = False
                        self.config_err = "'"+pkey+"' is not found in '%s'. Please check." % filename
                        return
                    else: # optional, assign default
                        self.setdefault(pkey, pdefault)
                else:
                    self.sectionfound.add(section)

        if "Linux" in self.sectionfound:
            print ("\n*** WARNING***\nSection [Linux] is obsolete. Please replace it by [Capture] in your .cfg files.\n")

        # Getting the stations parameters
        self.stations = []  # now defined as a list of dictionaries

        for i in range(self['number_of_stations']):
            section = "STATION_" + str(i+1)
            tmpDict = {}
            try:
                for parameter in (CALL_SIGN, FREQUENCY, COLOR):
                    tmpDict[parameter] = config_parser.get(section, parameter)
                self.stations.append(tmpDict)
            except ConfigParser.NoSectionError:
                self.config_ok = False
                self.config_err = section + "section is expected but missing from the config file."
                return
            except ConfigParser.NoOptionError:
                self.config_ok = False
                self.config_err = section + " does not have the 3 expected parameters in the config file. Please check."
                return
            else:
                self.sectionfound.add(section)

    def supersid_check(self):
        """Perform sanity checks when a .cfg file is read by 'supersid.py'.
        Verifies that all mandatory sections were read.
        Extend the keys with some other values for easier access."""
        if not self.config_ok: return

        for mandatory_section in ('PARAMETERS',):
            if mandatory_section not in self.sectionfound:
                self.config_ok = False
                self.config_err = mandatory_section + "section is mandatory but missing from the .cfg file."
                return

        # sanity check: as many Stations were read as announced by 'number_of_stations' (now section independent)
        if self['number_of_stations'] != len(self.stations):
            self.config_ok = False
            self.config_err = "'number_of_stations' does not match STATIONS found in supersid.cfg. Please check."
            return

        if 'stations' not in self:
            self[CALL_SIGN] = ",".join([s[CALL_SIGN] for s in self.stations])
            self[FREQUENCY] = ",".join([s[FREQUENCY] for s in self.stations])

        # log_type must be lower case and one of 'filtered' or 'raw'
        self['log_type'] = self['log_type'].lower()
        if self['log_type'] not in (FILTERED, RAW):
            self.config_ok = False
            self.config_err = "'log_type' must be either 'filtered' or 'raw' in supersid.cfg. Please check."
            return

        # 'hourly_save' must be UPPER CASE
        self['hourly_save'] = self['hourly_save'].upper()
        if self['hourly_save'] not in ('YES', 'NO'):
            self.config_ok = False
            self.config_err = "'hourly_save' must be either 'YES' or 'NO' in supersid.cfg. Please check."
            return

        # log_interval should be > 2
        if self['log_interval'] <= 2:
            self.config_ok = False
            self.config_err = "'log_interval' <= 2. Too fast! Please increase."
            return

        # check log_format
        self['log_format'] = self['log_format'].lower()
        if self['log_format'] not in (SID_FORMAT,SUPERSID_FORMAT, SUPERSID_EXTENDED, BOTH_EXTENDED):
            self.config_ok = False
            self.config_err = "'log_format' must be either 'sid_format' or 'supersid_format'/'supersid_extended'."
            return

        # Check the 'data_path' validity and create it as a Config instance property
        self.data_path = os.path.normpath(self['data_path'] or Config.DATA_PATH_NAME) + os.sep
        if not os.path.isdir(self.data_path):
            self.config_ok = False
            self.config_err = "'data_path' does not point to a valid directory:\n" + self.data_path
            return

        # default audio to pyaudio if not declared
        if "Audio" not in self:
            self["Audio"] = "pyaudio"

        # Just one choice: 'plot_offset = 0', for now ; not in the expected parameters list
        self['plot_offset'] = 0


if __name__ == '__main__':
    import sys
    # one argument: the .cfg file to read
    cfg = Config(sys.argv[1])
    cfg.supersid_check()
    if cfg.config_ok:
        print (cfg.filenames, "read successfully:")
    else:
        print ("Error:", cfg.config_err)
    for section in sorted(cfg.sectionfound):
        print ("-", section)
    print("-"*20)
    for k, v in sorted(cfg.items()):
        print (k,"=",v)
    print("-"*20)
    for st in cfg.stations:
        print (st)
